tasks/components.py

```diff
@@ -160,116 +161,217 @@ def has_type_component(content) -> bool:
+    "comp/otelcol/extension",
+    "comp/trace/agent",
+    "comp/core/workloadmeta",
 ]
 
-implementation_definitions = [
+# // TODO: (components), broken by new linter rules, fix in a follow-up PR
+ignore_package_name = [
+    "comp/trace/agent",
+    "comp/trace/compression",
+    "comp/core/workloadmeta",
+    "comp/core/healthprobe",
+    "comp/otelcol/collector",
+    "comp/otelcol/collector-contrib",
+    "comp/metadata/securityagent",
+    "comp/metadata/systemprobe",
+]
+
+mock_definitions = [
     "type Mock interface",
     "func Module() fxutil.Module",
     "func MockModule() fxutil.Module",
 ]
 
 
-def check_component_contents_and_file_hiearchy(entry_point):
-    # TODO: validate that def/component.go exists and defines `package <component>`
-    # TODO: validate that each of impl[-suffix]/[any].go defines `package <component>impl`
-    # TODO: validate that fx/fx.go, if it exists, defines `package <component>fx` or `package fx`
-    # TODO: validate that `def` is fx-free, and that `impl` is fx-free *except* for tests
-    content = entry_point.content
-    file = entry_point.file
-    directory = entry_point.dir
-
-    if not any(line.startswith('type Component interface') or line.startswith('type Component = ') for line in content):
-        return f"** {file} does not define a Component interface; skipping"
-
-    if str(file) in components_to_migrate:
-        return ""
+def check_component_contents_and_file_hiearchy(comp):
+    """
+    Check validity of component, returning first error, if any found
+    """
+    def_content = read_file_content(comp.def_file).split('\n')
+    root_path = pathlib.Path(comp.path)
+
+    # Definition file `def/component.go` must define a component interface
+    if not any(
+        line.startswith('type Component interface') or line.startswith('type Component = ') for line in def_content
+    ):
+        return f"** {comp.def_file} does not define a Component interface"
+
+    # Skip components that need to migrate
+    if str(comp.def_file) in components_to_migrate:
+        return
+
+    # Special case for api
+    if comp.def_file == 'comp/api/api/def/component.go':
+        return
+
+    # Definition file `def/component.go` must use `package <compname>`
+    pkgname = parse_package_name(comp.def_file)
+    if pkgname != comp.name:
+        if comp.path not in ignore_package_name:
+            return f"** {comp.def_file} has wrong package name '{pkgname}', must be '{comp.name}'"
+
+    # Definition file `def/component.go` must not contain a mock definition
+    for mock_definition in mock_definitions:
+        if any(line.startswith(mock_definition) for line in def_content):
+            return f"** {comp.def_file} defines '{mock_definition}' which should be in separate implementation. See docs/components/defining-components.md"
+
+    # Allowlist of components that do not use an implementation folder
+    if comp.path in components_missing_implementation_folder:
+        return
```

Comment:
Should we move this check above the previous ones? Since this check is the one checking for the original file structure. It seems weird that we do some check for `def/component.go` before it

Reply:
No, this isn't specific to v2 style components. The comments above are actually misleading, they are checking the `component.go` file for either situation, so I updated them.